snapshotter/demux/cache/snapshotter_cache.go

```diff
@@ -43,16 +43,18 @@ func (cache *SnapshotterCache) Get(ctx context.Context, key string, fetch Snapsh
 
 	if !ok {
 		cache.mutex.Lock()
-		defer cache.mutex.Unlock()
-
 		snapshotter, ok = cache.snapshotters[key]
+		cache.mutex.Unlock()
```

Comment:
we use a reader lock above. Should we be doing that here too?

Reply:
Oof that adds a new level of complexity if we need to account for a second double check lock.

Reply:
I'm confused. If `!ok`, wouldn't `cache.snapshotters[key]` be always nil?

Reply:
@kzys, unless another thread populates the cache after we have a cache miss but before we acquire the writers lock. The cache entry needs to be `.Close()` before it is garbage collected to cleanup system resources for metrics proxy.

Reply:
But @ginglis13 is correct this solution won't work for edge cases because we can leak now. So it requires `fetch` after reader's lock cache miss but before writer's lock acquisition.

Reply:
How about moving `cache.mutex.RUnlock()` in this if block? We wouldn't do much time consuming operations between RUnlock and Lock.

Reply:
That may result in deadlock. We'd block on Lock without releasing RUnlock. I don't believe RWMutex has brains to know assign writer's lock even if only one reader's lock is acquired by the same thread.